### ISSUE TITLE
Replace old Source.queue

### ISSUE THREAD

**Issue by patriknw** [akka/akka#29801]
In https://github.com/akka/akka/pull/29770 a new API was introduced for the dropNew use case.

We should deprecated the old Source.queue and add replacement for the overflow strategies that are still useful.
In my opinion we only need a replacement for the backpressure strategy.

For backpressure it would be useful to return a Future that is completed when the queue is ready to accept more. However, it's important that we say that it's not guaranteed to accept next offer just because the future was completed. There can be concurrent producers and all would be woken up by the Future. First offer wins (or as much as there is capacity) and others have to try again.

`fail` is already covered by the new API in https://github.com/akka/akka/pull/29770. Caller can look at the QueueOfferResult and call `BoundedSourceQueue.fail`.

I don't think there is a strong need for other strategies.

cc @jrudolph @seglo @raboof

**Comment by alex-ninja** [comment]
Hi @patriknw,

I'd like to start working on this issue, is that ok? Though I'll need some support because I'm not deeply familiar with Akka's internals.

I've gone through the ticket and the codebase. And at the moment I'd like to clarify a couple of questions:

1. The following method will be added to the `Source` (to let the user pass the overflow strategy):
```scala
def queue[T](bufferSize: Int, overflowStrategy: OverflowStrategy): Source[T, BoundedSourceQueue[T]]
```

but I guess it will be conflicting with the existing method:
```scala
def queue[T](bufferSize: Int, overflowStrategy: OverflowStrategy): Source[T, SourceQueueWithComplete[T]]
```

What is the best way to deal with the problem?

2. I am not really clear what exactly you're proposing by "it would be useful to return a Future". Do you want to change signature of the method `def offer(elem: T): QueueOfferResult` (it will break backward compatibility then) or to add an additional method `def offer(elem: T): Future[QueueOfferResult]` or to add an additional method that allows the caller to be notified when the queue has available capacity (smth like `def readyToAccept(): Future[Unit]`)?

PS:
Before making any changes, I'll write up a summary for the final approval.

**Comment by patriknw** [comment]
I don't think the `OverflowStrategy` parameter is useful here since it provides options that we will not support.

For the backpressure use case we could add a new:

```
def queueWithBackpressure[T](bufferSize: Int): Source[T, SourceQueueWithBackpressure[T]]
```

The `SourceQueueWithBackpressure` would provide:
```
def offer(elem: T): Future[Done]
```

Alternatively, the existing `BoundedSourceQueue` could add a new method:
```
def offerWithBackpressure(elem: T): Future[Done]
```

but I'm not sure how that would impact the efficiency of the implementation or if it would be confusing to combine offerWithBackpressure with the other offer?

**Comment by alex-ninja** [comment]
@patriknw 
I'm not yet sure what the best way to define the contract (re-use the existing or define a new queue interface) is, I think we can decide once we finalize the implementation approach. I have a few more questions on the implementation itself:
1. I feel we cannot guarantee that `offer` to the queue is always successful, therefore `Done` may not be the best fit. Do you propose return `Future` with a failure inside in case of an unsuccessful offer? I feel using `QueueOfferResult` might be a way to solve the problem. What do you think?
2. Do you assume using "async callback" approach for `Future` as it is done for `QueueSource`? Then we may experience the problem described in https://github.com/akka/akka/issues/29557.
3. As per my current understanding what you are proposing is a kind of simplified (without support for all overflow strategies) version of `QueueSource`. What is the conceptual difference between this new implementation and `QueueSource`?
4. On a separate note: as far as I understand, the idea of backpressure strategy is to ensure that producers do not push more data than downstream can consume. Why do we need to have `Future`? I believe futures have to be queued somewhere instead of blocking producers. I'm wondering why we can't just let producer be blocked on `offer` method if there is no free capacity in the queue. Could you please share your thoughts on this idea.

**Comment by patriknw** [comment]
I have a new suggestion. We could decouple the whenReady: Future[Done] from the offer return value. Adding a new whenReady method to BoundedSourceQueue that can be used for "waiting" until the queue is ready to accept more elements. Then the existing offer is used to emit the element.

My original description actually fits this approach even though I had a different implementation in mind at that point.

> For backpressure it would be useful to return a Future that is completed when the queue is ready to accept more. However, it's important that we say that it's not guaranteed to accept next offer just because the future was completed. There can be concurrent producers and all would be woken up by the Future. First offer wins (or as much as there is capacity) and others have to try again.

**Comment by alex-ninja** [comment]
@patriknw 
Hah! Actually that's exactly the way I thought it needs to be implemented initially, but then after a few rounds of reading the description I thought I was wrong. So your original description was fine, I was just confused by "replacement for the overflow strategies" phrase. 

I have raised a draft PR (without tests and documentation updates) with a few open questions. I'd be very grateful if you could help to resolve my doubts.

**Comment by jrudolph** [comment]
> For backpressure it would be useful to return a Future that is completed when the queue is ready to accept more. However, it's important that we say that it's not guaranteed to accept next offer just because the future was completed. There can be concurrent producers and all would be woken up by the Future. First offer wins (or as much as there is capacity) and others have to try again.

I'm not (yet) convinced of this idea. What are the potential use cases?

 * single producer: in that case it would somewhat work, however, since there are no watermarks etc to manage the buffer, in general it will be pulling for single elements. If we want a single Future-based pull mechanism we should just build that. In many cases, when the queue is full or almost full, we know that we near saturation of the downstream pipeline. Waking up the producer in such a case for single elements doesn't seem very efficient.
 * multiple producers: this is even more inefficient as we wake up `n` producers (and potentially threads) to contend for a single queue slot (or even less if there are also producers that just push without waiting for a signal).

In all cases, the question is why you would use a queue with a pulling mechanism in the first place, if you are able to backpressure the producer. Why not create a proper `Source` in these cases to use the inherent stream backpressure mechanism?

One answer might be that it looks convenient to be able to repeat an offer if the queue is currently full. But this would just create an inefficient ad-hoc buffer in front of the queue which is exactly what we want to avoid.

Maybe we can try to enumerate all the cases we'd like to cover and see if they can or should be solved by this queue or if there are alternatives or better solutions?

**Comment by jrudolph** [comment]
(Previous discussion in https://github.com/akka/akka/pull/29574#issuecomment-689547917.)

**Comment by patriknw** [comment]
> What are the potential use cases?

My primary reason is to be able to deprecate the old queue and have corresponding replacement. Previously we said that the new queue api covered everything meaningful except the backpressure strategy. This `whenReady` would be a way to cover that usage. It would be used when the producer can't just drop the rejected element. The producer would like to be woken up again when there is a chance that elements can be accepted again. It's likely that it will buffer (keep) the element while waiting, but it could use other ways of retrieving and emitting it again.

> In many cases, when the queue is full or almost full, we know that we near saturation of the downstream pipeline. Waking up the producer in such a case for single elements doesn't seem very efficient.

There are also cases when the queue isn't full, and elements can just be emitted efficiently without involving the whenReady wakup, but there must still be a way to handle the case when the queue is temporary full.

> this is even more inefficient as we wake up n producers

as mentioned in the PR it would be possible to wake up less or only one producer if we think that is important

> Why not create a proper Source in these cases to use the inherent stream backpressure mechanism?

Source.queue is used when interacting with something outside of streams, and it's not so nice if we say that users have to implement their own sources for all such integrations.

**Comment by patriknw** [comment]
>>this is even more inefficient as we wake up n producers

> as mentioned in the PR it would be possible to wake up less or only one producer if we think that is important

Another idea is to defer the wakup until half of the queue has been drained. For fairness we might want to reject new offers during the draining phase.

**Comment by jrudolph** [comment]
I'm still not convinced :)

All of these "solutions" sound like putting a queue in front of the queue. If you cannot drop elements, you can still use this queue by making it de-facto unbounded (size = Integer.MaxValue), if that's what you are after.

> > Why not create a proper Source in these cases to use the inherent stream backpressure mechanism?
> 
> Source.queue is used when interacting with something outside of streams, and it's not so nice if we say that users have to implement their own sources for all such integrations.

It's not a good use, though.
  * Either, the outside components is push-based, in which case no feedback mechanism is needed, either accept that some elements are dropped or use an unbounded queue.
  * Or, the outside interface is pull-based, in which case you can already use `Source.unfoldAsync`. If that API is too clunky, we can think of something better.

**Comment by patriknw** [comment]
So your suggestion is that we deprecate the old queue api and close this ticket (and the PR) without further action? Wait for feedback and take it from there.

**Comment by jrudolph** [comment]
> So your suggestion is that we deprecate the old queue api and close this ticket (and the PR) without further action? Wait for feedback and take it from there.

I think we should either take the time and consider options and then also update our recommendations in the docs and go through usages of `OverflowStrategy.backpressure` that we know about (e.g. in alpakka) and then make a decision. If we don't have the capacity right now, then postpone it. "Replace old Source.queue` is still a good goal but let's not just do something that we are not sure about.

**Comment by patriknw** [comment]
Agreed

**Comment by tg44** [comment]
Probably a bit off, but can't we move the buffer creation out as a parameter? [Here is a gist](https://gist.github.com/tg44/aa1d279b247d74e0ebca1489dc643410) with the implementation I started to use in one project. I think most of it could be useful in the mainline too (it started from a 2.5.23, so probably some of the things are not the same with the current version). My proposed changes;
 - Split the 0 size buffer implementation from the others. (Even to a new class.)
 - Give an option to the users to use a `() => Buffer[T]` function "constructor"
 - With the two modifications we could keep bincompat, and can add a new function to Source `def queue[T](queueCreator: () => Buffer[T], overflowStrategy: OverflowStrategy)`
 - If we move out the 0 size case, we can expose a new trait that gives back estimated usage information (used, capacity, isFull, isEmpty) for monitoring
 
 If you are interested in it, I propose a PR, but I found this thread, and probably I shouldn't work on this until I properly understand the state of the queue feature.

(The code shown there is for prioritized throttled email sending, for this, I needed a queue that can prioritize and I was really sad that I need to copy-paste from the original codebase. Also, we could handle the whole which to drop mechanism INSIDE the `Buffer[T]`, which would probably open some new possibilities. (Like the buffers enqueue method would return a class with 2 futures, passed to downstream or dropped, and can handle the overflow strategies, making the whole stage much cleaner.))

**Comment by hepin1989** [comment]
@tg44 I think your implementation is very useful, looking forward to the PR. we have something like that too, but instead of one priority queue, we have multiply queue associates with different priority.
